File: p1/src/application/ScreenViewCalculatorState.java
package application;

import java.math.BigDecimal;
import java.util.Optional;

public class ScreenViewCalculatorState extends BaseCalculatorState {


	protected ScreenViewCalculatorState(BigDecimal accumulator, BigDecimal screenValue,
			CalculatorOperator initialCalculatorOperator) {
		super(accumulator, screenValue, initialCalculatorOperator);
	}

	@Override
	public Calculator numericButtonPressed(int button) {
		if( thereIsPlaceForMoreDigitsOnScreen()){
		int numberOfDecimalPlaces = getNumberOfDecimalPlaces(screenValue);
			if (numberOfDecimalPlaces == 0) {
				screenValue = screenValue.multiply(new BigDecimal(10)).add(new BigDecimal(button));
			} else {
				BigDecimal thingToAdd = new BigDecimal("0.1").pow(numberOfDecimalPlaces + 1)
						.multiply(new BigDecimal(button));
				screenValue = screenValue.add(thingToAdd);
			}
			return new ScreenViewCalculatorState(accumulator, screenValue, currentCalculatorOperator);
		} else {
			return new ScreenViewCalculatorState(accumulator, screenValue, currentCalculatorOperator);
		}
	}
	
	private Optional<Calculator> executeOperation(CalculatorOperator operator){
		if( operator.equals(CalculatorOperator.DIVIDE) &&( screenValue.compareTo(BigDecimal.ZERO) == 0)){
			return Optional.of(new ErrorCalculatorState("DIV/0"));
		}
		accumulator = currentCalculatorOperator.executeAction(accumulator, screenValue);
		if( accumulatorIsTooBig() ){
			return Optional.of(new ErrorCalculatorState(bigDecimalToShortString(accumulator)+"E"));
		}
		return Optional.empty();
	}

	@Override
	public Calculator operatorButtonPressed(CalculatorOperator newOperator) {
		return executeOperation(newOperator)
				.orElse(new ScreenViewCalculatorState(accumulator, BigDecimal.ZERO, newOperator));	
	}

	@Override
	public Calculator dotButtonPressed() {
		return new DotPressedCalculatorState(accumulator, screenValue, currentCalculatorOperator);
	}

	@Override
	public Calculator clearButtonPressed() {
		return new ScreenViewCalculatorState(BigDecimal.ZERO, BigDecimal.ZERO, CalculatorOperator.ADD);
	}

	@Override
	public Calculator invertSignButtonPressed() {
		return new ScreenViewCalculatorState(accumulator, screenValue.negate(), currentCalculatorOperator);
	}

	@Override
	public Calculator squareRootButtonPressed() {
		return new ScreenViewCalculatorState(new BigDecimal(Math.sqrt(screenValue.doubleValue())), screenValue, currentCalculatorOperator);
	}

	@Override
	public Calculator equalButtonPressed() {
		return executeOperation(currentCalculatorOperator).orElse(
				new AccumulatorViewCalculatorState(accumulator, screenValue, currentCalculatorOperator));
	}

	@Override
	public String getDisplayedValue() {
		return bigDecimalToShortString(screenValue);
	}

}

File: NtrCalcWpf1/Calc1/Model/ScreenViewCalculatorState.cs
using Calc1.View1;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Calc1.Model
{
    class ScreenViewCalculatorState : BaseCalculatorState
    {
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        public ScreenViewCalculatorState(decimal accumulatorValue, string screenValue, CalculatorOperator calculatorOperator, int maxDigitsOnScreen) 
            : base(accumulatorValue, screenValue, calculatorOperator, maxDigitsOnScreen)
        {
                log.Debug("Nowy stan kalkulatora: ScreenViewCalculatorStateL Wartosc w akumulatorze:" + accumulatorValue + " Wartosc na ekranie:" + screenValue + " operator:" + calculatorOperator);
        }

        public override ICalculator numericButtonPressed(int button)
        {
            if (ScreenText.Length < maxDigitsOnScreen)
            {
                if (ScreenText.Equals("0"))
                {
                    ScreenValue = button;
                }
                else
                {
                    ScreenText = ScreenText + button;
                }
            }
			return new ScreenViewCalculatorState(AccumulatorValue, ScreenText, calculatorOperator, maxDigitsOnScreen);

        }
        public override ICalculator operatorButtonPressed(CalculatorOperator newCalculatorOperator)
        {
            log.Debug("Wcisnieto przycisk operatora " + calculatorOperator);
            try
            {
                if (calculatorOperator == CalculatorOperator.DIVIDE && ScreenValue == 0)
                {
                    return new ErrorCalculatorState(Constants.defaultErrorString, maxDigitsOnScreen);
                }
                AccumulatorValue = calculatorOperator.executeAction(ScreenValue, AccumulatorValue);
                return new ScreenViewCalculatorState(AccumulatorValue, "0", newCalculatorOperator, maxDigitsOnScreen);
            } catch( OverflowException e)
            {
                log.Warn("Overflow exception ", e);
                return new ErrorCalculatorState(Constants.OVERFLOW, maxDigitsOnScreen);
            }
        }
        public override ICalculator dotButtonPressed()
        {
            if (thereIsNoDot() && ScreenText.Length < maxDigitsOnScreen)
            {
                ScreenText = ScreenText + Constants.decimalNumberSeparator;
            }
            return this;
        }

        private bool thereIsNoDot(){
            return !ScreenText.Contains(Constants.decimalNumberSeparator);
        }

        public override ICalculator clearButtonPressed()
        {
            log.Debug("Wcisnieto przycisk czyszczenia ");
            return new ScreenViewCalculatorState(0, "0", CalculatorOperator.ADD, maxDigitsOnScreen);
        }
        public override ICalculator invertSignButtonPressed()
        {
            log.Debug("Wcisnieto przycisk odwrocenia znaku ");
            String newScreenValue = "?";
            if( ScreenText.StartsWith("-") ){
                newScreenValue = ScreenText.Substring(1);
            }
            else if (!ScreenText.Equals("0"))
            {
                newScreenValue = '-' + ScreenText;
            }
            else
            {
                newScreenValue = ScreenText;
            }
            return new ScreenViewCalculatorState(AccumulatorValue, newScreenValue, calculatorOperator, maxDigitsOnScreen);
        }

        public override ICalculator squareRootButtonPressed()
        {
            log.Debug("Wcisnieto przycisk pierwiastka ");
            if (ScreenValue < 0)
            {
                return new ErrorCalculatorState(Constants.defaultErrorString, maxDigitsOnScreen);
            }
            ScreenValue = (decimal)Math.Sqrt((double)ScreenValue);
            return new AccumulatorViewCalculatorState(ScreenValue, "0", CalculatorOperator.ADD, maxDigitsOnScreen);
        }
        public override ICalculator equalButtonPressed()
        {
            log.Debug("Wcisnieto przycisk wykonania rownania ");
            try
            {
                if (calculatorOperator == CalculatorOperator.DIVIDE && ScreenValue == 0)
                {
                    return new ErrorCalculatorState(Constants.defaultErrorString, maxDigitsOnScreen);
                }
                AccumulatorValue = calculatorOperator.executeAction(ScreenValue, AccumulatorValue);
                return new AccumulatorViewCalculatorState(AccumulatorValue, ScreenText, calculatorOperator, maxDigitsOnScreen);
            }
            catch (OverflowException e)
            {
                log.Warn("Overflow exception ", e);
                return new ErrorCalculatorState(Constants.OVERFLOW, maxDigitsOnScreen);
            }
        }
        public override ICalculator percentButtonPressed()
        {
            log.Debug("Wcisnieto przycisk procentu ");
            ScreenValue = AccumulatorValue * (ScreenValue * (decimal)0.01);
            if( calculatorOperator == CalculatorOperator.DIVIDE && ScreenValue == 0 ){
                return new ErrorCalculatorState(Constants.defaultErrorString, maxDigitsOnScreen);
            }
            AccumulatorValue = calculatorOperator.executeAction(ScreenValue, AccumulatorValue);
            return new AccumulatorViewCalculatorState(AccumulatorValue, "0", calculatorOperator, maxDigitsOnScreen);
        }
        public override CalculatorState CalculatorState
        {
            get
            {
                return CalculatorState.SCREEN_VIEW;
            }
        }
    }
}